File: examples/ts/src/utils/lending.ts
import {
  getLendingPositionAddress,
  getLendingVaultAddress,
  getTunaConfigAddress,
  LendingPosition as _LendingPosition,
  Vault as _Vault,
} from "@defituna/client";
import { Address, address, KeyPairSigner as _KeyPairSigner, KeyPairSigner, Rpc, SolanaRpcApi } from "@solana/kit";
import { findAssociatedTokenPda, TOKEN_PROGRAM_ADDRESS } from "@solana-program/token";
import { getMintDecimals, loadKeypair } from "src/utils/common";

/**
 * Prepares accounts and parameters for Tuna *lending pool* operations on the Solana blockchain.
 * @param rpc - The {@link Rpc RPC} client for Solana blockchain interactions.
 * @returns {Promise<{ tokenMintAddress: Address, nominalAmount: bigint, authority: KeyPairSigner, tunaConfigPda: Address, vaultPda: Address, lendingPositionPda: Address, authorityAta: Address, vaultAta: Address, decimalsScale: bigint }>}
 */
export async function prepareLendingAccountsAndParameters(rpc: Rpc<SolanaRpcApi>): Promise<{
  tokenMintAddress: Address;
  nominalAmount: bigint;
  authority: KeyPairSigner;
  tunaConfigPda: Address;
  vaultPda: Address;
  lendingPositionPda: Address;
  authorityAta: Address;
  vaultAta: Address;
  decimalsScale: bigint;
}> {
  /**
   * Define variables and accounts for Tuna *lending* operations;
   */
  /**
   * The {@link Address address} of the token mint to deposit/withdraw, identifying the target Tuna *Lending Vault*.
   * Set to the USDC token address in our examples;
   * There are methods in our sdk to fetch all available lending vaults and their respective mint addresses.
   */
  const tokenMintAddress: Address = address("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
  /**
   * The nominal amount to deposit, excluding *Token* decimals (e.g., 1 SOL as a flat value).
   * Note For deai
   */
  const nominalAmount = 1n;
  /**
   * The {@link _KeyPairSigner keypair} signing the transaction and owning the *Lending Position*,
   * defaults to the Solana config keypair (~/.config/solana/id.json).
   */
  const authority = await loadKeypair();

  /**
   * The {@link TunaConfig Tuna Config} Program Derived {@link Address Address} for Tuna operations, fetched from the Tuna Client.
   */
  const tunaConfigPda = (await getTunaConfigAddress())[0];
  /**
   * The {@link _Vault Lending Vault} Program Derived {@link Address address}, fetched from the Tuna Client.
   * Derived from the token mint address.
   */
  const vaultPda = (await getLendingVaultAddress(tokenMintAddress))[0];
  /**
   * The {@link _LendingPosition Lending Position} Program Derived {@link Address address}, fetched from the Tuna Client.
   * Derived from the authority (user) address and the token mint address.
   */
  const lendingPositionPda = (await getLendingPositionAddress(authority.address, tokenMintAddress))[0];

  /**
   * The Associated Token {@link Address Address}, owned by the *authority*.
   */
  const authorityAta = (
    await findAssociatedTokenPda({
      mint: tokenMintAddress,
      owner: authority.address,
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    })
  )[0];
  /**
   * The Associated Token {@link Address Address}, owned by the *Lending Vault*.
   */
  const vaultAta = (
    await findAssociatedTokenPda({
      mint: tokenMintAddress,
      owner: vaultPda,
      tokenProgram: TOKEN_PROGRAM_ADDRESS,
    })
  )[0];

  /**
   * Fetches token decimals for the Token, using {@link https://github.com/orca-so/whirlpools/tree/main/ts-sdk/client Orca's Whirlpool Client}.
   */
  const decimals = await getMintDecimals(rpc, tokenMintAddress);
  /**
   * The decimal scale to adjust nominal amounts for the Token based on its decimals.
   */
  const decimalsScale = BigInt(Math.pow(10, decimals));

  return {
    tokenMintAddress,
    nominalAmount,
    authority,
    tunaConfigPda,
    vaultPda,
    lendingPositionPda,
    authorityAta,
    vaultAta,
    decimalsScale,
  };
}

File: examples/rust/src/utils/lending.rs
use solana_client::rpc_client::RpcClient;
use solana_sdk::{pubkey::Pubkey, signer::Signer};
use spl_associated_token_account::get_associated_token_address;
use std::{error::Error, ops::Mul, str::FromStr};
use tuna_client::{get_lending_position_address, get_tuna_config_address, get_vault_address};

use super::common::get_mint_decimals;

pub struct LendingAccountsAndParameters {
  pub token_mint_address: Pubkey,
  pub amount: u64,
  pub tuna_config_pda: Pubkey,
  pub vault_pda: Pubkey,
  pub lending_position_pda: Pubkey,
  pub authority_ata: Pubkey,
  pub vault_ata: Pubkey,
}

pub fn prepare_lending_accounts_and_parameters(
  rpc: &RpcClient,
  authority: &Box<dyn Signer>,
) -> Result<LendingAccountsAndParameters, Box<dyn Error>> {
  let token_mint_address: Pubkey = Pubkey::from_str("So11111111111111111111111111111111111111112")?;

  let nominal_amount: f64 = 1.0;

  let decimals = get_mint_decimals(rpc, &token_mint_address)?;

  println!("decimals {}", decimals);

  let amount = nominal_amount
    .mul(f64::try_from(10_u32.pow(decimals as u32))?)
    .round()
    .abs() as u64;

  let (tuna_config_pda, _) = get_tuna_config_address();

  let (vault_pda, _) = get_vault_address(&token_mint_address);

  let (lending_position_pda, _) =
    get_lending_position_address(&authority.pubkey(), &token_mint_address);

  let authority_ata = get_associated_token_address(&authority.pubkey(), &token_mint_address);
  let vault_ata = get_associated_token_address(&vault_pda, &token_mint_address);

  Ok(LendingAccountsAndParameters {
    token_mint_address,
    amount,
    tuna_config_pda,
    vault_pda,
    lending_position_pda,
    authority_ata,
    vault_ata,
  })
}
